app/controllers/signup_controller.rb

```diff
@@ -35,36 +30,48 @@ def signup_form
   def signup_post
     handle_with(
       SignupForm,
+      # TODO: what is all this stuff getting set for?
       contracts_required: !contracts_not_required,
       client_app: get_client_app,
       is_bri_book: is_bri_book_adopter?,
       success: lambda {
-        save_unverified_user(@handler_result.outputs.user.id)
-        security_log(:user_viewed_signup_form, { user: @handler_result.outputs.user })
+        user = @handler_result.outputs.user
+        set_login_state(email: user.email_addresses.last,
+                        first_name: user.first_name,
+                        last_name: user.last_name,
+                        role: user.role)
+
+        security_log(:user_viewed_signup_form, { user: user })
         clear_cache_bri_marketing
+
+        user.faculty_status = 'needs_email_verified'
+        user.save!
+        sign_in!(user)
         redirect_to verify_email_by_pin_form_path and return
       },
       failure: lambda {
-        security_log(:sign_up_failed,
-                     { reason: @handler_result.errors.map(&:code), email: @handler_result.outputs.email })
-        render :signup_form and return
+        @errors = @handler_result.errors
+        security_log(:sign_up_failed, { reason: @handler_result.errors.map(&:code)})
+        render(:signup_form, params: { role: @selected_signup_role }) and return
       }
     )
   end
 
   def verify_email_by_pin_form
+    @signing_up_user = get_login_state
     render :email_verification_form
   end
 
   def verify_email_by_pin_post
+    @signing_up_user = get_login_state
     handle_with(
       VerifyEmailByPin,
-      email_address: unverified_user.email_addresses.first,
+      email_address: current_user.email_addresses.first,
```

Comment:
current_user can be nil or AnonymousUser here due to `skip_before_action :authenticate_user!`

Reply:
Still a problem, I think